e2e/imgbuild/imgbuild.go

```diff
@@ -154,6 +159,511 @@ func badPath(t *testing.T) {
+						{
+							Src: tmpfile.Name(),
+							Dst: "StageOne.txt",
+						},
+					},
+				},
+				{
+					Bootstrap: "docker",
+					From:      "alpine:latest",
+					Stage:     "two",
+					Files: []e2e.FilePair{
+						{
+							Src: tmpfile.Name(),
+							Dst: "StageTwo2.txt",
+						},
+						{
+							Src: tmpfile.Name(),
+							Dst: "StageTwo.txt",
+						},
+					},
+				},
+				{
+					Bootstrap: "docker",
+					From:      "alpine:latest",
+					Stage:     "three",
+					FilesFrom: []e2e.FileSection{
+						{
+							Stage: "one",
+							Files: []e2e.FilePair{
+								{
+									Src: "StageOne2.txt",
+									Dst: "StageOneCopy2.txt",
+								},
+								{
+									Src: "StageOne.txt",
+									Dst: "StageOneCopy.txt",
+								},
+							}},
+						{
+							Stage: "two",
+							Files: []e2e.FilePair{
+								{
+									Src: "StageTwo2.txt",
+									Dst: "StageTwoCopy2.txt",
+								},
+								{
+									Src: "StageTwo.txt",
+									Dst: "StageTwoCopy.txt",
+								},
+							},
+						}},
+				},
+				{
+					Bootstrap: "docker",
+					From:      "alpine:latest",
+					FilesFrom: []e2e.FileSection{
+						{
+							Stage: "three",
+							Files: []e2e.FilePair{
+								{
+									Src: "StageOneCopy2.txt",
+									Dst: "StageOneCopyFinal2.txt",
+								},
+								{
+									Src: "StageOneCopy.txt",
+									Dst: "StageOneCopyFinal.txt",
+								},
+								{
+									Src: "StageTwoCopy2.txt",
+									Dst: "StageTwoCopyFinal2.txt",
+								},
+								{
+									Src: "StageTwoCopy.txt",
+									Dst: "StageTwoCopyFinal.txt",
+								},
+							}}},
+				},
+			},
+			e2e.DefFileDetails{
+				Files: []e2e.FilePair{
+					{
+						Src: tmpfile.Name(),
+						Dst: "StageOneCopyFinal2.txt",
+					},
+					{
+						Src: tmpfile.Name(),
+						Dst: "StageOneCopyFinal.txt",
+					},
+					{
+						Src: tmpfile.Name(),
+						Dst: "StageTwoCopyFinal2.txt",
+					},
+					{
+						Src: tmpfile.Name(),
+						Dst: "StageTwoCopyFinal.txt",
+					},
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, test.WithPrivilege(func(t *testing.T) {
+
+			defFile := e2e.PrepareMultiStageDefFile(tt.dfd)
+			defer os.Remove(defFile)
+
+			opts := e2e.BuildOpts{
+				Sandbox: tt.sandbox,
+			}
+
+			imagePath := path.Join(testenv.TestDir, "container")
+			defer os.RemoveAll(imagePath)
+
+			if b, err := e2e.ImageBuild(testenv.CmdPath, opts, imagePath, defFile); err != nil {
+				t.Log(string(b))
+				t.Fatalf("unexpected failure: %v", err)
+			}
+
+			e2e.DefinitionImageVerify(t, testenv.CmdPath, imagePath, tt.correct)
+		}))
+	}
+
+}
+
+func buildDefinition(t *testing.T) {
+
+	tmpfile, err := ioutil.TempFile(testenv.TestDir, "testFile-")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	defer os.Remove(tmpfile.Name()) // clean up
+
+	if _, err := tmpfile.Write([]byte(testFileContent)); err != nil {
+		log.Fatal(err)
+	}
+	if err := tmpfile.Close(); err != nil {
+		log.Fatal(err)
+	}
+
+	tests := []struct {
```

Comment:
These are a bit easier to read in comparison to the multi-stage test cases, but I'm wondering if all the code to generate these templates is worth it over creating the definition files themselves and having them as test artifacts?

Reply:
I think it's worth it in this case because the amount of code we use to define what artifacts should be in the image is all we need to generate the definition itself from this struct. We could move to using definition files in a folder, but we would still need to have all these parameters defined in order to verify the final container was built properly.